core/src/main/java/org/elasticsearch/index/IndexSettings.java

```diff
@@ -108,6 +108,26 @@
         Setting.intSetting("index.max_script_fields", 32, 0, Property.Dynamic, Property.IndexScope);
 
     /**
+     * Index setting describing for NGramTokenizer and NGramTokenFilter
+     * the maximum difference between
+     * max_gram (maximum length of characters in a gram) and
+     * min_gram (minimum length of characters in a gram).
+     * The default value is 1 as this is default difference in NGramTokenizer,
+     * and is defensive as it prevents generating too many index terms.
+     */
+    public static final Setting<Integer> MAX_NGRAM_DIFF_SETTING =
+        Setting.intSetting("index.max_ngram_diff", 1, 0, Property.Dynamic, Property.IndexScope);
+
+    /**
+     * Index setting describing for ShingleTokenFilter
+     * the maximum difference between
+     * max_shingle_size and min_shingle_size.
+     * The default value is 3 is defensive as it prevents generating too many tokens.
+     */
```

Comment:
Yes too many tokens is problematic. I also don't like that we build invalid boolean queries when the diff is greater than 0. For instance with unigram set to true and shingles of size 2, the input `foo bar foobar` is analyzed as:
`position 1: (foo_bar, foo), position 2: (bar, bar_foobar) ...` which is problematic for any position query.
3 seems reasonable especially for fields that don't index positions.

Reply:
@jimczi  thanks for your comment.  I don't think I understand how diff in shingle size greater that 0   makes boolean queries invalid.   Does it mean that if we have tokens:  `{"token" :  "foo_bar", "position" : 1},  {"token" :  "foo", "position" : 1}, {"token" :  "bar", "position" : 2}, {"token" :  "bar_foobar", "position" : 2},`, we can't have a phrase query say: `"foo bar"`?

Reply:
`foo bar` would return the correct document but it would build an invalid phrase query:
```
"(foo_bar foo) bar"
```
... trying to find document with `foo_bar bar` as a phrase query which could be simplified in `foo_bar`. For boolean query it would not consider that `foo_bar` is enough to match `foo` AND `bar` so the bigram would be useless for matching this type of query. We have solutions on the query side to build the correct query (`foo_bar` OR (`foo` AND `bar`)) but this is currently deactivated because it can explode the number of clauses in the query if you have a large diff between min and max. 
For simplicity I prefer when a single shingle size is used for a shingle field. In this case a phrase query like `"foo bar foobar baz"` could be optimized to `"foo_bar foobar_baz"` with shingles of size 2 (`bar_foobar` can be safely ignored since we know that the previous match can only be `foo_bar`).